### PR TITLE
Adding a rule now focuses the newly created cell

### PR DESCRIPTION
**Pull request by azeghers**
__Which issue does this PR address?__

Related to [camunda/camunda-modeler#928](https://github.com/camunda/camunda-modeler/issues/928)

__Acceptance Criteria__

* [ ] Corresponds to the concept 
* [ ] Corresponds to the design 

__Definition of Done__

* [ ] corresponds to [the design principles](https://github.com/bpmn-io/design-principles)
* [ ] corresponds to [the code standards](https://github.com/bpmn-io/bpmn-js/blob/master/.github/CONTRIBUTING.md#creating-a-pull-request)
* [ ] passes Continuous Integration checks
* [ ] available as feature branch on GitHub
  * [ ] contains the cleaned up commit history
  * [ ] commit messages satisfy our [commit message guidelines](https://www.conventionalcommits.org/)
  * [ ] a single commit closes the issue via Closes #issuenr

### CODE REVIEW

**Comment by AZeghers** [comment]
Sorry about the double PR, I was having issues with the git history and the target branch :disappointed: 

I am still not entirely happy with the double event catch, but I haven't found another way to delay the creation of the element I'm trying to focus. I would be more than open to suggestions.
@barmac  You told me that a good test would be to make sure that the cell selection only happens once but I didn't figure out how to implement that test.

**Comment by pinussilvestrus** [comment]
One thing: the issue you referenced is not directly in this project, but in a downstream project (https://github.com/camunda/camunda-modeler/issues/928). I mostly reference it then via "Related to org/project#issue" or simply using the full URL of the issue. Also, let's use active phrases instead of passive ones.

So a better commit message could be

```text
feat(table): focus newly created cell on adding rule

Related to camunda/camunda-modeler#928
```

**Comment by AZeghers** [comment]
>The issue you referenced is not directly in this project...

Ah that's why it couldn't find the issue, of course.
<br />
  
>So a better commit message would be...

Do you want me to amend it?

**Comment by philippfromme** [comment]
You can mess around with the commits as much as you want so yes, amend it.

**Comment by barmac** [comment]
Also, you will probably need to update the PR description as well to prevent GitHub from prematurely closing the related issue when we merge the PR. We close downstream issues with dependency update. This means that the related issue should remain open until we install the update of dmn-js in Camunda Modeler repo.

**Comment by azeghers** [comment]
@barmac Only the PR description, not the commit message that states "Closes...", right?

**Comment by barmac** [comment]
> @barmac Only the PR description, not the commit message that states "Closes...", right?

Both :)

File: packages/dmn-js-decision-table/src/features/add-rule/PostAddRuleBehavior.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import CommandInterceptor from 'diagram-js/lib/command/CommandInterceptor';
+
+export default class PostAddRuleBehavior extends CommandInterceptor {
+  constructor(eventBus, selection) {
+    super(eventBus);
+
+    this.postExecuted('row.add', (event) => {
+
+      eventBus.once('elements.changed', 500, function() {
```
Comment (barmac) [inline packages/dmn-js-decision-table/src/features/add-rule/PostAddRuleBehavior.js:9]: Was there any particular reason to use arrow function in the first listener and then the `function` in the second?

Comment (azeghers) [reply packages/dmn-js-decision-table/src/features/add-rule/PostAddRuleBehavior.js:9]: There isn't, that's an oversight on my end

**Review by barmac — CHANGES_REQUESTED** [review]
The solution breaks the already existing feature when you press Enter with last row selected. Instead of selecting the cell below in the newly created row, we enforce selection of the first cell in the row. 

Screen Recording 2020-10-21 at 11 58 38

If you want, we can have a look into this together.

**Comment by azeghers** [comment]
Please note that clicking on the "add rule" row in the annotations column will focus on the first cell.
It's an edge case that had to be handled separately, because a generic implementation would crash as a side effect of #603.
It would probably be beneficial to treat the annotations cells as a part of a row, so their behavior can be unified.

File: packages/dmn-js-decision-table/src/features/add-rule/AddRule.js [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
 
-    eventBus.on('addRule', () => {
-      editorActions.trigger('addRule');
+    eventBus.on('addRule', (e, context) => {
+      const rule = editorActions.trigger('addRule');
+      const colIndex = context.colIndex || 0;
+      if (rule.cells[colIndex])
+        selection.select(rule.cells[colIndex]);
+      else
+        selection.select(rule.cells[0]);
```
[the remark this answers is not in the document]
Comment (azeghers) [reply packages/dmn-js-decision-table/src/features/add-rule/AddRule.js:13]: On it!

>I'm wondering whether we handle this via our linting.

My IDE should be following the linting rules and it doesn't tell me anything about clauses :man_shrugging:

File: packages/dmn-js-decision-table/test/spec/features/add-rule/AddRuleSpec.js [hunk 3]
```diff
@@ -66,6 +67,42 @@ describe('add input output', function() {
   }));
 
 
+  it('should add rule on click', inject(function(cellSelection, sheet) {
```
Comment (barmac) [inline packages/dmn-js-decision-table/test/spec/features/add-rule/AddRuleSpec.js:70]:
```suggestion
  it('should select cell in column which was clicked', inject(function(cellSelection, sheet) {
```

**Comment by pinussilvestrus** [comment]
Let's make sure we always have green commits, therefore not adding fixing-commits for previous commits in this PR, but simply squashing them. It keeps our history clean. Also, the component should be made clearer instead of just "table".

```text
feat(add-row): set focus to added rule

Related to camunda/camunda-modeler#928
```

**Review by pinussilvestrus — COMMENTED** [review]
works well for me 👍 

Kapture 2020-10-26 at 10 19 25

I was wondering why we don't keep the behavior into a separate behavior module. Is there a specific reason why it ended up in the general feature module? 

I don't mind if we keep it like that, just wanted to double-check.

**Comment by azeghers** [comment]
@pinussilvestrus If I understood correctly, the goal was to couple the 2 behaviors of adding a row and focusing the cell, instead of listening for a generic 'row.add'.
Also, adding rows and focusing the correct cells with the Enter key was already handled and additional event catching behavior would have lead to more unintended side effects and less readable code for _the next guy_.

**Review by pinussilvestrus — APPROVED** [review]

**Review by barmac — APPROVED** [review]

**Pull request merged**